### PR TITLE
feat: pass down indexed title from array field

### PR DESCRIPTION
**Pull request by Rozamo**
### Reasons for making this change

Fixes https://github.com/rjsf-team/react-jsonschema-form/issues/3983

## Changes
- Pass down indexed title from `ArrayField` to children
- Use the above `title` in `SchemaField`, `BooleanField` and `ObjectField` and `ArrayField`
- Add tests for the above changes

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [x] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Review by heath-freenome — CHANGES_REQUESTED** [review]
Can you please update the tests to validate these changes

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,12 @@ it according to semantic versioning. For example, if your PR adds a breaking cha
 should change the heading of the (upcoming) version to include a major version bump.
 
 -->
+# 5.16.1
```
[the remark this answers is not in the document]
Comment (Rozamo) [reply CHANGELOG.md]: Done.

**Comment by Rozamo** [comment]
> Can you please update the tests to validate these changes

Added tests for `ArrayField`. Let me know if tests at other places are also required.

**Review by nickgros — APPROVED** [review]

**Comment by nickgros** [comment]
My only concern is that we're changing the displayed index from 0-based to 1-based. Should we consider that a breaking change?

**Comment by Rozamo** [comment]
> My only concern is that we're changing the displayed index from 0-based to 1-based. Should we consider that a breaking change?

Very good point. Should I remove the `name` value from being used in the `fieldTitle` then? That way the if no title is provided in the array schema, the `name` with 0 based indexing would be used for the title as it was before. And if they provide the `title` they would get the 1 based indexed titles.

The only reason I used 1 based indexing for title was that I think they are mostly used for cosmetic purposes and it is more readable. If the dev is not providing the title it means that they might not require the readiblilty part.

**Review by heath-freenome — APPROVED** [review]

**Comment by heath-freenome** [comment]
> > My only concern is that we're changing the displayed index from 0-based to 1-based. Should we consider that a breaking change?
> 
> Very good point. Should I remove the `name` value from being used in the `fieldTitle` then? That way the if no title is provided in the array schema, the `name` with 0 based indexing would be used for the title as it was before. And if they provide the `title` they would get the 1 based indexed titles.
> 
> The only reason I used 1 based indexing for title was that I think they are mostly used for cosmetic purposes and it is more readable. If the dev is not providing the title it means that they might not require the readiblilty part.

Yes, having title be 1 based only is not a breaking change

**Pull request merged**